### ISSUE TITLE
[OnGoing] Inquiry on "The policy loss becomes NaN after training for over 100 steps"

### ISSUE THREAD

**Issue by Hsu5918** [Open-Reasoner-Zero/Open-Reasoner-Zero#28]
Very valuable work! However, I encountered a problem during training. Running orz_7b_ppo without modifying any parameters results in metrics fluctuating and not increasing during the first 100 steps. After 100 steps, the KL diverges, and the policy loss becomes NaN. What could be causing this issue?

**Comment by REIGN12** [comment]
Thank you for reporting this issue!
Can you please provide detailed training log, how you start the training, environment and the training curves? We'd love to reproduce your problem in our environment and hope to help better on your issue:-)

**Comment by Hsu5918** [comment]
Thank you for your response. After our investigation, the reason for the surge in KL divergence and the occurrence of NaN values is that the functions _broadcast_to_vllm_cudaipc or _broadcast_to_vllm were not called properly. This caused the VLLM model to continue using the base model for rollout, i.e., completely off-policy. This bug is likely due to the environment and the Hyperparameters colocate_all. 

1. We strongly recommend that anyone encountering similar issues **strictly follow the Dockerfile to ensure the environment is set up correctly**. For example, the author modified some source code of deepspeed package.
2. If you set allocate_all to false, then await self._sync_policy_weights_to_vllm() will not be executed.

**Comment by fiberleif** [comment]
> 1. We strongly recommend that anyone encountering similar issues **strictly follow the Dockerfile to ensure the environment is set up correctly**. For example, the author modified some source code of deepspeed package.

when l strictly follow the Dockerfile, there will be an error around flash-attention 2, so l am currently using torch==2.5.1, and flash-attn==2.7.0.post2, is it the same with your side? Thank you!

**Comment by Hsu5918** [comment]
We also encountered an error with flash-att, but we resolved it by downloading the installation package from https://github.com/Dao-AILab/flash-attention/releases and installing it locally. Note that you need to select the package based on your Torch version, CUDA version (you can choose a version lower than your current CUDA version), and Python version. Be sure to choose abiFALSE. We kept the final package version consistent with the requirements listed in https://github.com/user-attachments/files/18903173/requirements.txt.

**Comment by Hsu5918** [comment]
> > 1. We strongly recommend that anyone encountering similar issues **strictly follow the Dockerfile to ensure the environment is set up correctly**. For example, the author modified some source code of deepspeed package.
> 
> when l strictly follow the Dockerfile, there will be an error around flash-attention 2, so l am currently using torch==2.5.1, and flash-attn==2.7.0.post2, is it the same with your side? Thank you!

We also encountered an error with flash-att, but we resolved it by downloading the installation package from https://github.com/Dao-AILab/flash-attention/releases and installing it locally. Note that you need to select the package based on your Torch version, CUDA version (you can choose a version lower than your current CUDA version), and Python version. Be sure to choose abiFALSE. We kept the final package version consistent with the requirements listed in https://github.com/user-attachments/files/18903173/requirements.txt.

**Comment by fiberleif** [comment]
> > > 1. We strongly recommend that anyone encountering similar issues **strictly follow the Dockerfile to ensure the environment is set up correctly**. For example, the author modified some source code of deepspeed package.
> > 
> > 
> > when l strictly follow the Dockerfile, there will be an error around flash-attention 2, so l am currently using torch==2.5.1, and flash-attn==2.7.0.post2, is it the same with your side? Thank you!
> 
> We also encountered an error with flash-att, but we resolved it by downloading the installation package from https://github.com/Dao-AILab/flash-attention/releases and installing it locally. Note that you need to select the package based on your Torch version, CUDA version (you can choose a version lower than your current CUDA version), and Python version. Be sure to choose abiFALSE. We kept the final package version consistent with the requirements listed in https://github.com/user-attachments/files/18903173/requirements.txt.

Thank you for sharing these insights!

**Comment by liangz1995** [comment]
> Thank you for your response. After our investigation, the reason for the surge in KL divergence and the occurrence of NaN values is that the functions _broadcast_to_vllm_cudaipc or _broadcast_to_vllm were not called properly. This caused the VLLM model to continue using the base model for rollout, i.e., completely off-policy. This bug is likely due to the environment and the Hyperparameters colocate_all.
> 
> 1. We strongly recommend that anyone encountering similar issues **strictly follow the Dockerfile to ensure the environment is set up correctly**. For example, the author modified some source code of deepspeed package.
> 2. If you set allocate_all to false, then await self._sync_policy_weights_to_vllm() will not be executed.

I meet this problem too, I follow the dockerfile to build my image but still encounter the KL divergence surge problem when setting collocate_all=False. Also delved into  the source code, I found that when collocate_all is False, function _sync_policy_weights_to_vllm will not be called except the begining

**Comment by mrsempress** [comment]
Since I cannot use docker, I use [link](https://github.com/user-attachments/files/18903173/requirements.txt) from @Hsu5918 to install it. But some packages in it are local, such as `torchdata @ file:///opt/pytorch/data`, so I did not install them. I hope @Hsu5918 can provide the installation method of these local packages. 

After update the repo in [commits](https://github.com/Open-Reasoner-Zero/Open-Reasoner-Zero/commit/e008f6d95f0b9a0e992f6b8bac912515b50a4634), I also encountered the situation that kl is Nan after several iterations. Before that, the value of kl was more than 100.
![Image](https://github.com/user-attachments/assets/88c92c17-49c0-4741-90ba-4b327835075d)

If the author has modified the deepspeed package like Hsu5918 guess, can you tell me the specific location?
